python/pyspark/pandas/frame.py

```diff
@@ -7648,52 +7654,6 @@ def to_list(os: Optional[Union[Name, List[Name]]]) -> List[Label]:
             else:
                 return [o if is_name_like_tuple(o) else (o,) for o in os]
 
-        if isinstance(right, ps.Series):
```

Comment:
Can we minimize the codes moving around? It's difficult to review. You can make a separate PR to refactor this first.

Reply:
Sure I can do this. FYI, unfortunately, as part of this refactor implementation a large chunk of code was required to be indented one level, which will still cause a large diff block, even though the changes are fairly minimal.

How would you like me to structure the PRs? Would you rather:
(1) create a isolated code repositioning branch that will be merged to master in one PR, and then a second join-feature update branch that will be separately merged to master in a second PR? 
Or (2) create a code repositioning branch, with a PR to master, and then another join-feature update branch with a PR to the code repositioning branch. The join feature branch will get merged into the other branch first, then that branch to master.

I think (2) makes more sense, since if we don't agree on the implementation of the join feature, then the code repositioning will just be unnecessarily merged in noise.

Reply:
Actually, I would prefer to create an isolated PR that refactors this into, for example, a separate private function or nested function as an example.

I think it's fine even without merging the join feature - this logic you are moving seems better be separate and factored out in any event.

Reply:
@HyukjinKwon - I just pushed an update to address cleaning up the logic in the merge function (the first half of the function, up until the creation of the `joined_table`). And actually, I hope you don't mind, but I've also included the cross and conditional merge logic into the PR as well, because it actually is only a minimal marginal addition to the cleanup refactor to implement both features.

If we only want the PR to address cleaning up the function (i.e. no behavior change/new features), the only  change we would need to make to the PR is to reduce this block of code:
https://github.com/apache/spark/blob/f052ff0c90de487342a2da71a56cdee2b1b3099f/python/pyspark/pandas/frame.py#L7720-L7737

To this subset of that block of code:
```
left_key_names, right_key_names = self._resolve_merge_key_names(
    right, on, left_on, right_on, left_index, right_index
)
join_condition = _resolve_merge_join_condition_from_key_names(
    left_table, right_table, left_key_names, right_key_names
)
```

I think there will still be a bit to digest in the PR, but it's all related to the null-op cleanup refactor, after that's been processed the new feature additions are very simple to follow.

I think this would be a potentially good way to move forward, let me know your thoughts on this approach.

Reply:
@HyukjinKwon - I just pushed an update to address cleaning up the logic in the merge function (the first half of the function, up until the creation of the `joined_table`). And actually, I hope you don't mind, but I've also included the cross and conditional merge logic into the PR as well, because it actually is only a minimal marginal addition to the cleanup refactor to implement both features. (This is the reason I kept the changes in this PR rather than a new PR)

If we only want the PR to address cleaning up the function (i.e. no behavior change/new features), the only  change we would need to make to the PR is to reduce this block of code:
https://github.com/apache/spark/blob/f052ff0c90de487342a2da71a56cdee2b1b3099f/python/pyspark/pandas/frame.py#L7720-L7737

To this subset of that block of code above:
https://github.com/apache/spark/blob/f052ff0c90de487342a2da71a56cdee2b1b3099f/python/pyspark/pandas/frame.py#L7731-L7737

I think there will still be a bit to digest in the PR, but it's all related to the null-op cleanup refactor, after that's been processed the new feature additions (from the first snippet above) are very simple to follow. I've added some comments to explain the cleanup reorg part.

I think this would be a good way to move forward, let me know your thoughts on this approach.